includes/di.php

```diff
@@ -0,0 +1,67 @@
+<?php
+/**
+ * DI config.
+ *
+ * @package   Google\Site_Kit
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit;
+
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\DI\DI_Container;
+use Google\Site_Kit\Core\Storage\Transients;
+use Google\Site_Kit\Core\Storage\User_Options;
+
+/**
+ * Sets up a new DI container and returns it.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @return DI_Container DI container.
+ */
+function setup_di_container() {
+	$di = new DI_Container();
+
+	// Define scalar entities.
+	$di['MAIN_FILE'] = GOOGLESITEKIT_PLUGIN_MAIN_FILE;
+
+	// Define service entities.
+	$di['context'] = function( $di ) {
+		return new Context( $di['MAIN_FILE'] );
+	};
+
+	$di->set_services(
+		array(
+			'plugin'      => '\\Google\\Site_Kit\\Plugin',
+			'modules'     => '\\Google\\Site_Kit\\Core\\Modules\\Modules',
+			'assets'      => '\\Google\\Site_Kit\\Core\\Assets\\Assets',
+			'options'     => '\\Google\\Site_Kit\\Core\\Storage\\Options',
+			'transients'  => '\\Google\\Site_Kit\\Core\\Storage\\Transients',
+			'admin_bar'   => '\\Google\\Site_Kit\\Core\\Admin_Bar\\Admin_Bar',
+			'rest_routes' => '\\Google\\Site_Kit\\Core\\REST_API\\REST_Routes',
+		)
+	);
+
+	$di['user_options'] = function( $di ) {
+		return new User_Options( $di['context'] );
+	};
+
+	$di['authentication'] = function( $di ) {
+		return new Authentication( $di['context'], $di['options'], $di['user_options'], $di['transients'] );
+	};
```

Comment:
See my first comment, this looks quite intransparent, e.g. why is one passing a value, another one passing a callback, another one using `set_services` with class names?

If this was all using methods like `set_value` and `set_service`, it would be easier to understand. Also this shows our `set_service` should support passing the entire callback, so that those dependencies can be filled in. Passing the class name should be supported too, but that is only enough for the "simple" services without constructor dependencies.

Reply:
> this looks quite intransparent, e.g. why is one passing a value, another one passing a callback, another one using set_services with class names?

@felixarntz this is just a demonstration of how some services that require a more complex setup can be defined in the DI container. Eventually, after refactoring, the `Authentication` class will also be added to the `set_services` definitions.

> Also this shows our set_service should support passing the entire callback, so that those dependencies can be filled in. Passing the class name should be supported too, but that is only enough for the "simple" services without constructor dependencies.

The idea for the `set_service` method was exactly to allow us to define "simple" services that can be easily instantiated without passing anything into the constructor. Perhaps we should rename that method to `set_clean_services`?

Reply:
> Eventually, after refactoring, the `Authentication` class will also be added to the `set_services` definitions.

Thanks, makes sense.

> The idea for the `set_service` method was exactly to allow us to define "simple" services that can be easily instantiated without passing anything into the constructor. Perhaps we should rename that method to `set_clean_services`?

I don't think this differentiation makes sense. Why does it matter whether a service requires constructor parameters or not? Either way it's a service, so it should be set through `set_service` (or the `set_services` wrapper). Why not support `string|callable` (class name or create function) for that and handle it accordingly?

Reply:
Ok, updated the `set_service` method to accept class names as well :+1: